### DISCUSSION TITLE
Second-level query cache - entity always fetches lazy loaded data

### DISCUSSION THREAD

**Discussion by nfplee** [nhibernate/nhibernate-core#3175] [Q&A]
I have an issue when I wish to retrieve an entity (which has already been cached) but this time I fetch a lazy loaded property it will trigger a hit to database. For example say I have the following models:

```cs
public class ContentItem {
    public virtual int Id { get; set; }
    public virtual string Name { get; set; }
    public virtual Head? Head { get; set; }
}

public class Head {
    public virtual string Title { get; set; }
}
```

Which are mapped like so:

```cs
public class ContentItemMapping : ClassMapping<ContentItem> {
    public ContentItemMapping() {
        Table("ContentItems");
        Id(x => x.Id, m => m.Generator(Generators.Identity));
        Property(x => x.Name);
        Component(x => x.Head);
        Cache(x => x.Usage(CacheUsage.ReadWrite));
    }
}

public class HeadMapping : ComponentMapping<Head> {
    public HeadMapping() {
        Property(x => x.Title);
        Lazy(true);
    }
}
```

Here's my sample application, indicating each time the database is hit:

```cs
public static async Task Main() {
    using (var sessionFactory = CreateSessionFactory()) {
        using (var session = CreateSession(sessionFactory)) {
            var contentItems = await GetAllContentItemsAsync(session); // Hits the database.

            var contentItem = await GetContentItemAsync(session); // Hits the database.
            var title = contentItem?.Head?.Title;
        }

        using (var session = CreateSession(sessionFactory)) {
            var contentItem = await GetContentItemAsync(session);
            var title = contentItem?.Head?.Title; // Hits the database.
        }

        using (var session = CreateSession(sessionFactory)) {
            var contentItem = await GetContentItemAsync(session);
            var title = contentItem?.Head?.Title; // Hits the database.
        }
    }
}

private static async Task<IList<ContentItem>> GetAllContentItemsAsync(ISession session) {
    return await session.Query<ContentItem>()
        .WithOptions(o => o.SetCacheable(true))
        .ToListAsync();
}

private static async Task<ContentItem> GetContentItemAsync(ISession session) {
    return await session.Query<ContentItem>()
        .WithOptions(o => o.SetCacheable(true))
        .Fetch(i => i.Head)
        .FirstOrDefaultAsync();
}
```

Notice when I try to retrieve an entity via `GetContentItemAsync` (which has already been cached by `GetAllContentItemsAsync`), the first time it does a fresh hit to the database (to fetch the lazy loaded data as requested) but the cache is not updated. Now any further calls to the lazy loaded data will also trigger a hit to the database since I assume was because it was not included in the original query which cached the entity.

I've been playing around with the cache mode but each one seems trigger a fresh hit to the database which defeats the purpose of a cache.

I was originally writing this as a bug but then I looked at the docs and realized that the query cache just caches the id's and the entity cache is responsible for caching the values.

I'd appreciate if someone could show me a way around this as I can imagine this is quite a common scenario. For example if you have paged list of blog posts which are cached and then you click through to the details page which contains more details, you wouldn't want to load all of the data on the list page.

**Comment by bahusoid** [comment]
Try removing `.Fetch(i => i.Head)` from your query and let `Head` be lazy loaded by first access. And check if it's retrieved from cache on next access.

**Comment by nfplee** [comment]
Thanks but removing the fetch performed slightly worse as now it adds an extra hit to the database when it first calls the `Head.Title` property. Here's a sample application for you to see it in action:

[ConsoleApp1.zip](https://github.com/nhibernate/nhibernate-core/files/9697671/ConsoleApp1.zip)

**Comment by bahusoid** [comment]
I see it's been discussed here https://github.com/nhibernate/nhibernate-core/pull/1922#issuecomment-448643931

And if I understand correctly  lazy property initialization should be propagated to cache  (at least lazy loaded case after https://github.com/nhibernate/nhibernate-core/pull/1949)

Seems @maca88 is the best person to shed the light on it...

**Comment by fredericDelaporte** [comment]
The cache disables itself quite fast when you use it without transactions. Please wrap all your session usage inside transactions, to make sure the lack of transaction is not the issue. See [here](https://nhibernate.info/doc/nhibernate-reference/performance.html#performance-cache):

> The second level cache requires the use of transactions, be it through transaction scopes or NHibernate transactions. Interacting with the data store without an explicit transaction is discouraged, and will not allow the second level cache to work as intended.

**Comment by nfplee** [comment]
Thanks @fredericDelaporte but wrapping the calls in a transaction didn't seem to make much difference. Here's my updated sample application:

[ConsoleApp1.zip](https://github.com/nhibernate/nhibernate-core/files/9701547/ConsoleApp1.zip)

Looking at the changes in the https://github.com/nhibernate/nhibernate-core/pull/1949 pull request (by @maca88) I should be able to create a test case for this if you feel it's a bug?
